File: src/api/impl/branch.py
# -*- mode: python; encoding: utf-8 -*-
#
# Copyright 2014 the Critic contributors, Opera Software ASA
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may not
# use this file except in compliance with the License.  You may obtain a copy of
# the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
# License for the specific language governing permissions and limitations under
# the License.

import api
import apiobject

class Branch(apiobject.APIObject):
    wrapper_class = api.branch.Branch

    def __init__(self, branch_id, name, repository_id, head_id):
        self.id = branch_id
        self.name = name
        self.__repository_id = repository_id
        self.__head_id = head_id
        self.__head = None
        self.__commits = None

    def getRepository(self, critic):
        return api.repository.fetch(critic, repository_id=self.__repository_id)

    def getHead(self, critic):
        if self.__head is None:
            self.__head = api.commit.fetch(
                self.getRepository(critic), commit_id=self.__head_id)
        return self.__head

    def getCommits(self, critic):
        if self.__commits is None:
            repository = self.getRepository(critic)
            cursor = critic.getDatabaseCursor()
            cursor.execute("""SELECT commit
                                FROM branchcommits
                               WHERE branch=%s""",
                           (self.id,))
            self.__commits = api.commitset.create(
                critic, (api.commit.fetch(repository, commit_id)
                         for (commit_id,) in cursor))
        return self.__commits

@Branch.cached()
def fetch(critic, branch_id, repository, name):
    cursor = critic.getDatabaseCursor()
    if branch_id is not None:
        cursor.execute("""SELECT id, name, repository, head
                            FROM branches
                           WHERE id=%s""",
                       (branch_id,))
    else:
        cursor.execute("""SELECT id, name, repository, head
                            FROM branches
                           WHERE repository=%s
                             AND name=%s""",
                       (repository.id, name,))
    try:
        return next(Branch.make(critic, cursor))
    except StopIteration:
        if branch_id is not None:
            raise api.branch.InvalidBranchId(branch_id)
        else:
            raise api.branch.InvalidBranchName(name)

def fetchAll(critic, repository):
    cursor = critic.getDatabaseCursor()
    if repository is not None:
        cursor.execute("""SELECT id, name, repository, head
                            FROM branches
                           WHERE repository=%s
                        ORDER BY name""",
                       (repository.id,))
    else:
        cursor.execute("""SELECT id, name, repository, head
                            FROM branches
                        ORDER BY name""")
    return list(Branch.make(critic, cursor))

File: ui/src/actions/branch.ts
/*
 * Copyright 2017 the Critic contributors, Opera Software ASA
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License.  You may obtain a copy
 * of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

import {
  fetch,
  include,
  withArgument,
  withContext,
  withParameters,
} from "../resources"
import { AsyncThunk, Dispatch, Thunk } from "../state"
import {
  BRANCH_COMMITS_UPDATE,
  SET_CREATED_BRANCHES,
  SET_UPDATED_BRANCHES,
  SET_RECENT_BRANCHES,
  Action,
} from "."
import {
  BranchID,
  CommitID,
  UserID,
  RepositoryID,
  RebaseID,
} from "../resources/types"
import { RequestParams } from "../utils/Fetch.types"
import Branch from "../resources/branch"
import { PaginationInfo } from "../resources/fetch"
import { assertNotNull } from "../debug"
import { PaginationThunk } from "../utils/Pagination"
import Commit from "../resources/commit"

export const setCreatedBranches = (branchIDs: BranchID[]): Action => ({
  type: SET_CREATED_BRANCHES,
  branchIDs,
})

export const setUpdatedBranches = (branchIDs: BranchID[]): Action => ({
  type: SET_UPDATED_BRANCHES,
  branchIDs,
})

type BranchCommitsUpdateKey = {
  branchID: number
  afterRebaseID?: number | null
  offset?: number | null
  count?: number | null
}
export const branchCommitsUpdate = (
  {
    branchID,
    afterRebaseID = null,
    offset = null,
    count = null,
  }: BranchCommitsUpdateKey,
  commitIDs: CommitID[],
): Action => ({
  type: BRANCH_COMMITS_UPDATE,
  branchID,
  afterRebaseID,
  offset,
  count,
  commitIDs,
})

export const setRecentBranches = (
  repositoryID: number,
  offset: number,
  count: number,
  branchIDs: BranchID[],
): Action => ({
  type: SET_RECENT_BRANCHES,
  repositoryID,
  offset,
  count,
  branchIDs,
})

export const loadCreated = (
  userID: number,
  count = 5,
): AsyncThunk<void> => async (dispatch) => {
  const { primary } = await dispatch(
    fetch(
      "branches",
      withParameters({
        created_by: userID,
        exclude_reviewed_branches: "yes",
        count,
      }),
      include("repositories"),
    ),
  )
  if (primary) dispatch(setCreatedBranches(primary.map((branch) => branch.id)))
}

const loadUpdated = (userID: number, count = 5): AsyncThunk<void> => async (
  dispatch,
) => {
  const { primary } = await dispatch(
    fetch(
      "branches",
      withParameters({
        updated_by: userID,
        exclude_reviewed_branches: "yes",
        count,
      }),
    ),
  )
  if (primary) dispatch(setUpdatedBranches(primary.map((branch) => branch.id)))
}

export const loadBranchesForDashboard = (userID: UserID): Thunk<void> => (
  dispatch,
) => {
  dispatch(loadCreated(userID))
  dispatch(loadUpdated(userID))
}

type BranchByNameParams = { repositoryID: RepositoryID; name: string }

export const loadBranch = (branchID: BranchID) =>
  fetch("branches", withArgument(branchID))

export const loadBranchByName = (repository: RepositoryID, name: string) =>
  fetch("branches", withParameters({ repository, name }))

type BranchCommitsOptions = {
  afterRebaseID?: number | null
  offset?: number
  count?: number
}

export const loadBranchCommits = (
  branchID: BranchID,
  offset: number,
  count: number,
): PaginationThunk<Commit> => async (dispatch) => {
  const { primary, pagination } = await dispatch(
    fetch(
      "commits",
      withContext("branches", branchID),
      withParameters({ offset, count }),
    ),
  )
  assertNotNull(pagination)
  return [primary, pagination]
}

export const loadBranchCommitsAfterRebase = (
  branchID: BranchID,
  afterRebaseID: RebaseID,
): AsyncThunk<Commit[]> => async (dispatch) => {
  const { primary } = await dispatch(
    fetch(
      "commits",
      withContext("branches", branchID),
      withParameters({ after_rebase: afterRebaseID }),
    ),
  )
  if (primary) {
    dispatch(
      branchCommitsUpdate(
        { branchID, afterRebaseID },
        primary.map((commit) => commit.id),
      ),
    )
  }
  return primary
}

export const loadRecentBranches = (
  repositoryID: RepositoryID,
  {
    offset = 0,
    count = 10,
  }: {
    offset?: number
    count?: number
  },
) => async (dispatch: Dispatch) => {
  const { primary } = await dispatch(
    fetch("branches", {
      params: { repository: repositoryID, order_by: "updated", offset, count },
      include: ["commits", "branches"],
    }),
  )
  if (primary) {
    const branchIDs = primary.map((branch) => branch.id)
    dispatch(setRecentBranches(repositoryID, offset, count, branchIDs))
  }
}

export const loadRepositoryBranches = (
  repositoryID: RepositoryID,
  offset: number,
  count: number,
): AsyncThunk<[Branch[], PaginationInfo]> => async (dispatch) => {
  const { primary, pagination } = await dispatch(
    fetch(
      "branches",
      withContext("repositories", repositoryID),
      withParameters({ offset, count }),
    ),
  )
  assertNotNull(pagination)
  return [primary, pagination]
}
